src/coreclr/tools/aot/ILCompiler.MetadataTransform/ILCompiler/Metadata/Transform.CustomAttribute.cs

```diff
@@ -20,195 +19,245 @@ private List<CustomAttribute> HandleCustomAttributes(Cts.Ecma.EcmaModule module,
+                Cts.TypeDesc enumType = Cts.CustomAttributeTypeNameParser.GetTypeByCustomAttributeTypeName(module, valueReader.ReadSerializedString());
+                return new ConstantBoxedEnumValue
+                {
+                    Value = HandleCustomAttributeConstantValue(module, TypeDescToSerializationTypeCode(enumType), ref valueReader),
+                    Type = HandleType(enumType)
+                };
             }
 
-            Debug.Assert(value is Cts.TypeDesc);
-            Debug.Assert(type is Cts.MetadataType
-                && ((Cts.MetadataType)type).Name == "Type"
-                && ((Cts.MetadataType)type).Namespace == "System");
+            if (typeCode == Ecma.SerializationTypeCode.String)
+            {
+                string s = valueReader.ReadSerializedString();
+                return s == null ? new ConstantReferenceValue() : HandleString(s);
+            }
+
+            if (typeCode == Ecma.SerializationTypeCode.Type)
+            {
+                string s = valueReader.ReadSerializedString();
+                return s == null ? new ConstantReferenceValue() : HandleType(Cts.CustomAttributeTypeNameParser.GetTypeByCustomAttributeTypeName(module, s));
+            }
 
-            return HandleType((Cts.TypeDesc)value);
+            return typeCode switch
+            {
+                Ecma.SerializationTypeCode.Boolean => new ConstantBooleanValue { Value = valueReader.ReadBoolean() },
+                Ecma.SerializationTypeCode.Char => new ConstantCharValue { Value = valueReader.ReadChar() },
+                Ecma.SerializationTypeCode.Byte => new ConstantByteValue { Value = valueReader.ReadByte() },
+                Ecma.SerializationTypeCode.SByte => new ConstantSByteValue { Value = valueReader.ReadSByte() },
+                Ecma.SerializationTypeCode.Int16 => new ConstantInt16Value { Value = valueReader.ReadInt16() },
+                Ecma.SerializationTypeCode.UInt16 => new ConstantUInt16Value { Value = valueReader.ReadUInt16() },
+                Ecma.SerializationTypeCode.Int32 => new ConstantInt32Value { Value = valueReader.ReadInt32() },
+                Ecma.SerializationTypeCode.UInt32 => new ConstantUInt32Value { Value = valueReader.ReadUInt32() },
+                Ecma.SerializationTypeCode.Int64 => new ConstantInt64Value { Value = valueReader.ReadInt64() },
+                Ecma.SerializationTypeCode.UInt64 => new ConstantUInt64Value { Value = valueReader.ReadUInt64() },
+                Ecma.SerializationTypeCode.Single => new ConstantSingleValue { Value = valueReader.ReadSingle() },
+                Ecma.SerializationTypeCode.Double => new ConstantDoubleValue { Value = valueReader.ReadDouble() },
+                Ecma.SerializationTypeCode.SZArray => HandleCustomAttributeConstantArray(module, valueReader.ReadSerializationTypeCode(), ref valueReader),
+                _ => throw new System.Exception()
```

Comment:
`Cts.ThrowHelper.ThrowBadImageFormatException` or `throw new UnreachableException()` ?

Reply:
`Cts.ThrowHelper.ThrowBadImageFormatException` is not a throw for C#, so it won't work.

Any opinion about simply disabling warning CS8509 ILCompiler-project-wide?

We keep having to work around it by adding artificial throws, or worse, by adding `_ =>` for something that could be one concrete value (`SomeFoo =>`), but the right-hand side happens to throw for all the other possibilities, so it's as good as an extra `_ => throw new Blah()` line.

Reply:
> adding _ => for something that could be one concrete value (SomeFoo =>)

The compiler is going to add throw for cases that are not covered by the switch. The difference here is whether the code is explicitly handling it with a throw or whether you let the compiler to generate the throw implicitly. Many of the compiler warnings and our own coding conventions rules force you to write more verbose and explicit code, so I see the CS8509 warning as just another one of those.

I think we should throw BIF for invalid input files when we have a choice. If these default switch cases are reachable for invalid input (it is not obvious to me whether it is the case), we may want to add `Cts.ThrowHelper.CreateBadImageFormatException` and change them to `throw Cts.ThrowHelper.CreateBadImageFormatException()`.

Reply:
The thing is that I would be fine with the `SwitchExpressionException` the compiler throws. Anything that kills the compilation is fine by me.

I agree that our coding standards err on the more verbose side, so `throw new UnrechableException` sounds like the best fix.

Throwing a Cts.BadImageException wouldn't make a difference here. This spot is non-recoverable; we're in the middle of generating metadata and can't change our mind about it. We already inspected the attribute for dependencies during dependency analysis so at this point we know it doesn't refer to something that doesn't exist (we'd skip it if it does). If it has some other structural issue we didn't detect during dependency analysis, I'm fine with a compiler crash. We try to be resilient for invalid inputs that could be the result of bad assembly version management, but we don't try to be resilient against source-to-IL compiler bugs. Too much resilience makes it difficult to surface _our own_ bugs because this would also paper over those.

Reply:
> We try to be resilient for invalid inputs that could be the result of bad assembly version management,

Parsing of custom attribute blobs can hit BadImageFormat exception due to user error. If the custom attribute references enum from a different assembly and the enum underlying type changes due to user error, parsing of the custom attribute blob will get lost and it will likely end up hitting one of these unreachable cases.

Reply:
That's a good point but it still needs to go to the other spot that parses it first (at the time we analyze dependencies). #101164 has the fix.